docs/proposals/admission_control_changes.md

```diff
@@ -0,0 +1,485 @@
+  4. e.Storage.Get
+  1. rest.BeforeDelete
+  1. e.Storage.Delete
+  1. e.AfterDelete
+  1. e.Decorator
+
+#### Proposed behavior
+
+1. decode request (DeleteOptions)
+2. r.Get
+1. r.BeforeDelete
+  2. If strategy-based, call rest.BeforeDelete
+  1. Otherwise, just do whatever is in the rest.Deleter
+1. admit.Validate
+1. admit.EnforceConstraints
+1. r.Delete
+  2. For strategy-based handlers:
+    1. e.Storage.Delete
+    1. e.AfterDelete
+    1. e.Decorator
+2. admit.Finish
+
+### DeleteCollection
+
+#### Current behavior
+
+1. admit.Admit
+2. decode ListOptions from request parameters
+3. decode request body (DeleteOptions)
+4. r.DeleteCollection (the items below are part of the generic registry store / strategy-based storage, `pkg/registry/generic/registry/store.go`)
+  5. e.List
+  1. run *n* worker goroutines in parallel:
+    1. e.Delete (same as Delete -> Current behavior -> 3)
+
+#### Proposed behavior
+
+1. admit.Validate
+2. admit.EnforceConstraints
+2. decode ListOptions from request parameters
+3. decode request body (DeleteOptions)
+4. r.DeleteCollection ...
+5. admit.Finish
+
+### Connect
+
+#### Current behavior
+
+1. connecter.NewConnectOptions
+2. decode request connect options
+3. admit.Admit
+4. connecter.Connect().ServeHTTP()
+
+#### Proposed behavior
+
+1. connecter.NewConnectOptions
+2. decode request connect options
+1. admit.Validate
+2. connecter.Connect().ServeHTTP()
+
+## Ability to run out-of-tree plugins
+
+Admission plugins are all currently compiled into the apiserver binary. If you
+want to run custom plugins, you have to modify the Kubernetes source code and
+recompile the apiserver. Some examples of custom plugins that an administrator
+might want to run (these are all actual plugins in OpenShift today):
+
+- default restart-never and restart-on-failure pods' `activeDeadlineSeconds`
+  based on either a namespace-level annotation, or the plugin's configured value
+  (although this probably should be part of LimitRange).
+- apply quota at the cluster level instead of per-namespace
+- assign default resource request values relative to user-specified limits for
+  pods to support an administrator-defined overcommit target
+- restrict who is allowed to set the `nodeName` or `nodeSelector` on pods
+- ensure you can't exec or attach to a pod with a privileged security context if
+  you don't have privileges to create a privileged pod
+- supply default values for certain fields on Build Pods when unset (e.g.
+  environment variables)
+- enforce cluster-wide configuration values on all Build Pods (i.e.
+  override user-supplied data if necessary)
+- automatically provision a Jenkins server in a user's namespace if it doesn't
+  exist when a pipeline BuildConfig is created
+
+We would like to have the ability to run a single external plugin at the end of
+each admission phase. This will enable cluster administrators to inject custom
+admission logic into an existing kube-apiserver without having to recompile it.
```

Comment:
what's the proposed mechanism for this?  the external admission control plugin is part of an external api server and k8s is configured to invoke that api server?  (part of the admission controller plugin configuration provided to k8s defines the external admission controllers to invoke, including host, CA, etc?)

and if so, perhaps that configuration can tell k8s which external admission controllers should be fail open vs fail closed (and/or for which namespaces).